base/card-api.gts

```diff
@@ -1266,49 +1308,66 @@ export async function recompute(card: Card, opts?: RecomputeOptions): Promise<vo
+  let deserialized = getDataBucket(instance);
+  let identityContext = identityContexts.get(instance) ?? new IdentityContext();
+  let maybeStale = deserialized.get(fieldName as string);
+  if (isStaleValue(maybeStale)) {
+    let field = getField(Reflect.getPrototypeOf(instance)!.constructor as typeof Card, fieldName as string);
+    if (!field) {
+      throw new Error(`the field '${fieldName as string} does not exist in card ${instance.constructor.name}'`);
+    }
+    let { computeVia } = field;
+    if (!computeVia) {
+      throw new Error(`the field '${fieldName as string}' is not a computed field in card ${instance.constructor.name}`);
+    }
+    compute = typeof computeVia === 'function' ? computeVia.bind(instance) : () => (instance as any)[computeVia as string]();
+  }
+
+  try {
+    result = await compute();
+  } catch (e: any) {
+    if (isNotLoadedError(e)) {
+      // taking advantage of the identityMap regardless of whether loadFields is set
+      let fieldValue = identityContext.identities.get(e.reference);
+      if (fieldValue) {
+        deserialized.set(fieldName as string, fieldValue);
+        return fieldValue as T[K];
+      }
 
-        if (opts?.loadFields) {
-          let card = Reflect.getPrototypeOf(model)!.constructor as typeof Card;
-          let field = getField(card, fieldName as string);
-          if (!field) {
-            throw new Error(`the field '${fieldName as string} does not exist in card ${card.name}'`);
-          }
-          let instance = await loadMissingField(card, field, e, identityContext);
-          deserialized.set(fieldName as string, instance);
-        } else {
-          isLoaded = true;
-        }
-      } else if (isNotReadyError(e)) {
-        let { model: depModel, computeVia, fieldName: depField } = e;
-        if (typeof computeVia === 'function') {
-          deserialized.set(depField, await computeVia.bind(depModel)());
-        } else {
-          deserialized.set(depField, await depModel[computeVia]());
+      if (opts?.loadFields) {
+        let card = Reflect.getPrototypeOf(instance)!.constructor as typeof Card;
+        let field = getField(card, fieldName as string);
+        if (!field) {
+          throw new Error(`the field '${fieldName as string} does not exist in card ${card.name}'`);
         }
-      } else {
-        throw e;
+        let fieldValue = await loadMissingField(card, field, e, identityContext);
+        deserialized.set(fieldName as string, fieldValue);
+        result = fieldValue as T[K];
       }
+      return result;
+    } else if (isNotReadyError(e)) {
+      let { model: depModel, computeVia, fieldName: depField } = e;
+      let nestedCompute = typeof computeVia === 'function' ? computeVia.bind(depModel) : () => depModel[computeVia as string]();
+      let value = await getIfReady(depModel, depField, nestedCompute, opts);
+      if (isNotReadyValue(value)) {
+        return value;
+      }
+      deserialized.set(depField, value);
```

Comment:
i think this is unnecessary--we'll set this in the recurse into `getIfReady` above